libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApiTest.kt

```diff
@@ -695,39 +696,52 @@ class MapboxRouteLineApiTest {
-                        6
-                    )
-                    every { distanceTraveled } returns 0f
-                    every { step } returns mockk {
-                        every { distance() } returns route.legs()!![0].steps()!![2].distance()
+    fun updateTraveledRouteLine_pointUpdateIntervalRespected() =
+        coroutineRule.runBlockingTest {
+            val options = MapboxRouteLineOptions.Builder(ctx)
+                .withVanishingRouteLineEnabled(true)
+                .displayRestrictedRoadSections(false)
+                .vanishingRouteLineUpdateInterval(TimeUnit.MILLISECONDS.toNanos(1200))
+                .build()
+            val api = MapboxRouteLineApi(options)
+            val route = getRoute()
+            val lineString = LineString.fromPolyline(route.geometry() ?: "", Constants.PRECISION_6)
+            val routeProgress = mockk<RouteProgress> {
+                every { currentLegProgress } returns mockk {
+                    every { legIndex } returns 0
+                    every { currentStepProgress } returns mockk {
+                        every { stepPoints } returns PolylineUtils.decode(
+                            route.legs()!![0].steps()!![2].geometry()!!,
+                            6
+                        )
+                        every { distanceTraveled } returns 0f
+                        every { step } returns mockk {
+                            every { distance() } returns route.legs()!![0].steps()!![2].distance()
+                        }
+                        every { stepIndex } returns 2
                     }
-                    every { stepIndex } returns 2
                 }
             }
-        }
 
-        api.updateVanishingPointState(RouteProgressState.TRACKING)
-        api.setRoutes(listOf(RouteLine(route, null)))
-        api.updateUpcomingRoutePointIndex(routeProgress)
+            api.updateVanishingPointState(RouteProgressState.TRACKING)
+            api.setRoutes(listOf(RouteLine(route, null)))
+            api.updateUpcomingRoutePointIndex(routeProgress)
 
-        val result = api.updateTraveledRouteLine(lineString.coordinates()[1])
+            pauseDispatcher {
+                val result1 = api.updateTraveledRouteLine(lineString.coordinates()[1])
+                assertTrue(result1.isValue)
 
-        assertTrue(result.isError)
-    }
+                Thread.sleep(1000L)
+                api.updateUpcomingRoutePointIndex(routeProgress) // only update the progress
+                Thread.sleep(300L) // in summary we've waited for 1.3s since last point update
+                val result2 = api.updateTraveledRouteLine(lineString.coordinates()[1])
+                assertTrue(result2.isValue) // should succeed because threshold was 1.2s
+
+                Thread.sleep(500L) // wait less than threshold
+                val result3 = api.updateTraveledRouteLine(lineString.coordinates()[1])
+                assertTrue(result3.isError)
```

Comment:
I would hope to use a more elegant way of advancing time than `Thread.sleep` but since we need to impact the machine's elapsed time, this is the easiest way I could find to create this one test. If there were more use cases of this, we could consider adding a custom "time provider" class where we could mock and control the elapsed time. Open to suggestions though.

Reply:
We can also directly mock `System.nanoTime`.

Reply:
Not if you call `System.nanoTime()` directly.

You could try using mockk to fake `nanoTime()` responses between `updateTraveledRouteLine` calls.
```kotlin
mockkStatic(System::nanoTime)
every { System.nanoTime() } returns 1
// or
every { System.nanoTime() } returnsMany 1 andThen 2 andThen 3
```

Or, abstract calls to System.nanoTime().

```kotlin
// introduce new interface
interface SystemClock {
    fun nanoTime(): Long
}

// add new field to MapboxRouteLIneApi class
@VisibleForTesting
internal var systemClock: SystemClock = object : SystemClock {
    override fun nanoTime() = System.nanoTime()
}
// replace `System.nanoTime()` usage with `systemClock.nanoTime()`

// and override its implementation in tests
val clock = mockk<SystemClock>()
evert { clock.nanoTime() } returns 1 

api.systemClock = clock
api.updateTraveledRouteLine(lineString.coordinates()[1])
```

Reply:
Right, I think @Zayankovsky had the former in mind, and I had the latter in mind :+1: Both could work, we can revisit this as we go.